hazelcast/src/main/java/com/hazelcast/map/impl/MapContainer.java

```diff
@@ -158,7 +159,24 @@ public Indexes createIndexes(boolean global) {
                 .indexProvider(mapServiceContext.getIndexProvider(mapConfig))
                 .usesCachedQueryableEntries(mapConfig.getCacheDeserializedValues() != CacheDeserializedValues.NEVER)
                 .partitionCount(partitionCount)
-                .build();
+                .resultFilter(queryableEntry -> hasNotExpired(queryableEntry)).build();
+    }
+
+    /**
+     * @return {@code true} if queryableEntry has
+     * not expired, otherwise returns {@code false}
+     */
+    private boolean hasNotExpired(QueryableEntry queryableEntry) {
```

Comment:
Is this method called on the partition thread?

Reply:
yes, when it is not global index or i can't find counter example at least. I have added check `assertRunningOnPartitionThread` when index is not global.